test/qbwc/integration/session_test.rb

```diff
@@ -136,4 +136,111 @@ def requests(job, session, data)
       assert timothy_session.ticket != margaret_session.ticket
     end
   end
+
+  test "two sessions that advance to the next request_index should not clobber each other" do
+    # In pseudocode, we are doing this:
+    #   Advance Tim's request_index, mocking a delay between the SELECT and the UPDATE inside advance_next_request
+    #   Advance Margaret's request index between Tim's SELECT and his UPDATE
+    #   This would cause Margaret's update to effectively be rolled back because Tim ends up saving the wrong value for Margaret
+    #
+    # The solution is to perform the select and update inside of a transaction with locking,
+    # ensuring that the record that is SELECTed doesn't change before its UPDATEd.
+    # This forces Margaret to wait until Tim has finished his operation before she can lock the row for her update.
+
+    QBWC.add_job(:session_test_1, true, COMPANY, ConditionalTestWorker)
+
+    timothy_session  = QBWC::Session.new("timothy", COMPANY)
+    margaret_session = QBWC::Session.new("margaret", COMPANY)
+
+    job = QBWC.jobs.first
+
+    delayed_save_for_tim = lambda do
+      @@sleep_1_for_timothy_and_0_for_margaret ||= 1
+      the_delay = @@sleep_1_for_timothy_and_0_for_margaret
+      @@sleep_1_for_timothy_and_0_for_margaret = 0
+      sleep(the_delay)
+
+      self.send("__minitest_any_instance_stub__save")
+    end
+
+    QBWC::ActiveRecord::Job::QbwcJob.stub_any_instance(:save, delayed_save_for_tim) do
+      threads = []
+      threads << Thread.new {
+        # This would not blow up in mysql (or any other real DBMS),
+        # but since we're using sqlite3 to test, we look for an explosion
+        error = assert_raises(ActiveRecord::StatementInvalid) do
+          job.advance_next_request(timothy_session)
+        end
+        assert_match /SQLite3::BusyException/, error.message
+      }
+
+      threads << Thread.new {
+        sleep(0.25)
+        # This would not blow up in mysql (or any other real DBMS),
+        # but since we're using sqlite3 to test, we look for an explosion
+        error = assert_raises(ActiveRecord::StatementInvalid) do
+          job.advance_next_request(margaret_session)
+        end
+        assert_match /SQLite3::BusyException/, error.message
+      }
+
+      threads.each { |thread| thread.join }
+    end
```

Comment:
I am not entirely sold on the use of `sleep` to manage the concurrency

Specifically I would recommend that instead of trying once to "shoot the gap" to test the locking mechanism, instead we should prepare a loop that fights against the lock more than 100 times or so in quick succession to better ensure we have entered the race condition case